src/palimpzest/operators/physical.py

```diff
@@ -27,24 +26,46 @@ class PhysicalOp:
     synthesizedFns = {}
     solver = Solver(verbose=LOG_LLM_OUTPUT)
 
-    def __init__(self, outputSchema: Schema, shouldProfile = False) -> None:
+    def __init__(self, outputSchema: Schema, source: PhysicalOp=None, shouldProfile=False) -> None:
         self.outputSchema = outputSchema
+        self.source = source
         self.datadir = DataDirectory()
-
         self.shouldProfile = shouldProfile
-        self.profiler = Profiler()
-        self.profile = self.profiler.iter_profiler
+
+        # NOTE: this must be overridden in each physical operator's __init__ method;
+        #       we have to do it their b/c the opId() (which is an argument to the
+        #       profiler's constructor) may not be valid until the physical operator
+        #       has initialized all of its member fields
+        self.profiler = None
 
     def opId(self) -> str:
         raise NotImplementedError("Abstract method")
 
     def dumpPhysicalTree(self) -> Tuple[PhysicalOp, Union[PhysicalOp, None]]:
-        raise NotImplementedError("Abstract method")
+        """Return the physical tree of operators."""
+        if self.source is None:
+            return (self, None)
 
-    def getProfilingData(self) -> Dict[str, Any]:
-        raise NotImplementedError("Abstract method")
+        return (self, self.source.dumpPhysicalTree())
 
-    def estimateCost(self) -> Dict[str, Any]:
+    def getProfilingData(self) -> OperatorStats:
+        # simply return stats for this operator if there is no source
+        if self.shouldProfile and self.source is None:
+            return self.profiler.get_data()
+
+        # otherwise, fetch the source operator's stats first, and then return
+        # the current operator's stats w/a copy of its sources' stats
+        elif self.shouldProfile:
+            source_operator_stats = self.source.getProfilingData()
+            operator_stats = self.profiler.get_data()
+            operator_stats.source_op_stats = source_operator_stats
```

Comment:
Again, should we enforce this by requiring that source_operator_stats be passed into profiler.get_data()?

Reply:
Longer term yes, although I think this will complicate a few things to change it now, so let's mark this as a tech debt to resolve post-preprint